core/templates/components/question-directives/questions-list/questions-list.component.ts

```diff
@@ -282,59 +282,55 @@ export class QuestionsListComponent implements OnInit, OnDestroy {
     }
 
     let numberOfQuestions = this.questionSummariesForOneSkill.length;
-    await this.skillBackendApiService.checkSkillAssignmentForDiagnosticTest(
-      this.selectedSkillId).then((response) => {
-      if (response && (
-        numberOfQuestions <=
-          AppConstants.MINIMUM_QUESTION_COUNT_FOR_A_DIAGNOSTIC_TEST_SKILL)) {
-        this.alertsService.addInfoMessage(
-          'The skill must be removed from the diagnostic test first.');
-        questionDeletionIsAllowed = false;
-      }
-    });
+    await this.skillBackendApiService
+      .getTopicNamesWithGivenSkillAssignedForDiagnosticTest(
+        this.selectedSkillId).then((topicNames) => {
+        if ((topicNames.length > 0) && (
+          numberOfQuestions <=
+            AppConstants.MINIMUM_QUESTION_COUNT_FOR_A_DIAGNOSTIC_TEST_SKILL)) {
+          let alertMessge = (
+            'This skill is used as the only diagnostic test skill in the ' +
+            'following topics: ' + topicNames.join(', ') + '. Please remove ' +
+            'this skill from the diagnostic test of those topics first'
+          );
+          this.alertsService.addInfoMessage(alertMessge, 7000);
+          questionDeletionIsAllowed = false;
+        }
+      });
     return questionDeletionIsAllowed;
   }
 
+  editQuestionSkillLinks(questionId: string, skillId: string): void {
+    this.editableQuestionBackendApiService.editQuestionSkillLinksAsync(
+      questionId, [
+        {
+          id: skillId,
+          task: 'remove'
+        } as SkillLinkageModificationsArray
+      ]
+    ).then(() => {
+      this.questionsListService.resetPageNumber();
+      this.questionsListService.getQuestionSummariesAsync(
+        this.selectedSkillId, true, true);
+      this.alertsService.addSuccessMessage('Deleted Question');
+      this._removeArrayElement(questionId);
+    });
+  }
+
   deleteQuestionFromSkill(
       questionId: string, skillDescription: string): void {
-    this.questionDeletionCheck().then((response) => {
+    this.isQuestionDeletionAllowed().then((response) => {
       if (!response) {
         return;
       }
+
       this.deletedQuestionIds.push(questionId);
-      // For the case when, it is in the skill editor.
       if (this.allSkillSummaries.length === 0) {
-        this.editableQuestionBackendApiService.editQuestionSkillLinksAsync(
-          questionId, [
-            {
-              id: this.selectedSkillId,
-              task: 'remove'
-            } as SkillLinkageModificationsArray
-          ]
-        ).then(() => {
-          this.questionsListService.resetPageNumber();
-          this.questionsListService.getQuestionSummariesAsync(
-            this.selectedSkillId, true, true);
-          this.alertsService.addSuccessMessage('Deleted Question');
-          this._removeArrayElement(questionId);
-        });
+        this.editQuestionSkillLinks(questionId, this.selectedSkillId);
       } else {
         this.allSkillSummaries.forEach((summary) => {
```

Comment:
Should we break after finding the summary we want?

Reply:
updated with the use of the `find` method.